src/Orleans.Core.Abstractions/IDs/StreamId.cs

```diff
@@ -0,0 +1,155 @@
+
+        public static StreamId Create(IStreamIdentity streamIdentity) => Create(streamIdentity.Namespace, streamIdentity.Guid);
+
+        public int CompareTo(StreamId other) => FullKey.Span.SequenceCompareTo(other.FullKey.Span);
+
+        public bool Equals(StreamId other) => FullKey.Span.SequenceEqual(other.FullKey.Span);
+
+        public void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            info.AddValue("fk", FullKey.ToArray());
+            info.AddValue("ki", this.keyIndex);
+            info.AddValue("fh", this.hash);
+        }
+
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+
+            if (!Namespace.IsEmpty)
+                sb.Append(Encoding.UTF8.GetString(Namespace.ToArray()));
+            else
+                sb.Append("null");
+
+            sb.Append($"/{Encoding.UTF8.GetString(Key.ToArray())}");
+
+            return sb.ToString();
+        }
+
+        public override int GetHashCode() => this.hash;
+    }
+
+    [Immutable]
+    [Serializable]
+    [StructLayout(LayoutKind.Auto)]
+    internal readonly struct InternalStreamId : IEquatable<InternalStreamId>, IComparable<InternalStreamId>, ISerializable
+    {
+        public string ProviderName { get; }
+
+        public StreamId StreamId { get; }
+
+        public InternalStreamId(string providerName, StreamId streamId)
+        {
+            ProviderName = providerName;
+            StreamId = streamId;
+        }
+
+        private InternalStreamId(SerializationInfo info, StreamingContext context)
+        {
+            ProviderName = (string) info.GetValue("pvn", typeof(string));
+            StreamId = (StreamId) info.GetValue("sid", typeof(StreamId));
+        }
+
+        public static implicit operator StreamId(InternalStreamId internalStreamId) => internalStreamId.StreamId;
+
+        public bool Equals(InternalStreamId other) => StreamId.Equals(other) && ProviderName.Equals(other.ProviderName);
+
+        public int CompareTo(InternalStreamId other) => throw new NotImplementedException();
+
+        public void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            info.AddValue("pvn", ProviderName);
+            info.AddValue("sid", StreamId, typeof(StreamId));
+        }
+
+        public override int GetHashCode() => HashCode.Combine(ProviderName, StreamId);
+
+        public override string ToString()
+        {
+            return $"{ProviderName}/{StreamId.ToString()}";
+        }
+    }
+
+    public static class StreamIdExtensions
+    {
+        public static string GetKeyAsString(this StreamId streamId) => Encoding.UTF8.GetString(streamId.Key.ToArray());
+
+        public static string GetNamespace(this StreamId streamId)
+        {
+            if (streamId.Namespace.IsEmpty)
+                return null;
+
+            return Encoding.UTF8.GetString(streamId.Namespace.ToArray());
```

Comment:
If this is called frequently, we can avoid the extra array allocation with some `#if NETCOREAPP` code